### ISSUE TITLE
support for user-defined networks?

### ISSUE THREAD

**Issue by jayarjo** [testcontainers/testcontainers-node#66]
Any way/plans to support user-defined networks? From what I see `testcontainers-java` has [such functionality](https://www.testcontainers.org/features/networking#advanced-networking).

**Comment by cristianrgreco** [comment]
Yes, it's in the README: `.withNetworkMode("network_name")`

**Comment by cristianrgreco** [comment]
Sorry, I closed the issue by mistake.

**Comment by jayarjo** [comment]
But you can't create the network itself with it, or can you? I thought it was only for well the mode (e.g. bridge, etc)

**Comment by cristianrgreco** [comment]
That's right, currently you cannot create the network with testcontainers. Feel free to raise a PR, if not I'll look into this when I can.

**Comment by katerina-semikina** [comment]
Hi everyone. Looking forward for this useful change. How the things are going?

**Comment by jayarjo** [comment]
> Hi everyone. Looking forward for this useful change. How the things are going?

Вау, девчонка интересуется возможностью создания сетки между докер контейнерами! Клааасс! Я тебе обещаю, если ответишь на этот комментарий я добавлю детали о моем ПиаРе в ридми ) До сих пор никак руки не доходили просто...

**Comment by katerina-semikina** [comment]
> > Hi everyone. Looking forward for this useful change. How the things are going?
> 
> Вау, девчонка интересуется возможностью создания сетки между докер контейнерами! Клааасс! Я тебе обещаю, если ответишь на этот комментарий я добавлю детали о моем ПиаРе в ридми ) До сих пор никак руки не доходили просто...

Это было бы здорово!

**Comment by jayarjo** [comment]
> Это было бы здорово!

Сделано 🙌  
😉

**Comment by cristianrgreco** [comment]
Hi @jayarjo, I've pushed a couple of minor changes, do you want to take a look? If OK I'll release it

**Comment by jayarjo** [comment]
@cristianrgreco well... I'd say not ok, since I took those interfaces and method names from Testcontainers Java, shouldn't we follow them in that?

Also the idea behind `Network.newNetwork()` was that network creation is asynchronous,  so `new Network()` doesn't properly represent the process. Also I didn't want Network instance to be accessible/available before `Network.newNetwork()` resolves. That way you have natural defense in place against people trying to invoke methods on the instance before it's ready to be used. Now that you've changed that, you need to provide some alternatives, like state of the Network maybe or something.

In fact you are using similar pattern for Container instantiation logic, so why break it?

**Comment by cristianrgreco** [comment]
Seems like a big issue considering I changed a static creator method to a constructor, I even said on the PR that I would make these changes once merged and you agreed :slightly_smiling_face: 

> since I took those interfaces and method names from Testcontainers Java, shouldn't we follow them in that?

Yes but within reason, we should leverage the benefits of JS. 

This is a reason why I got rid of the static creator function, because there's no compiler stopping someone from invoking `new Network()` anyway, unlike in the Java port where the constructor can be made private. It also means we don't have to write extra code and to test at runtime that we throw an error if the constructor is being called, what's the point? 

By moving the creation options into the constructor, the network instance can be re-used many times:

```javascript
const network = new Network({ ... lots of options })
const network1 = await network.start();
const network2 = await network.start();
```

> In fact you are using similar pattern for Container instantiation logic, so why break it for Networks?

This is the pattern we use for instantiaing containers, you'll find it's the same as the change I made:

```javascript
const container = await new GenericContainer().start();
```

**Comment by jayarjo** [comment]
First of all you are maintainer of this package, so you decide :)

But here are my points:

>  I even said on the PR that I would make these changes once merged and you agreed

I thought you were talking about examples in README, 'cause you brought this topic previously I remember and I had very similar reaction.

> Yes but within reason, we should leverage the benefits of JS.

Since big part of documentation is on Testcontainers Java side, I'd expect us following them as close as possible. 

> there's no compiler stopping someone from invoking new Network() anyway

Network is not exported, only it's interface is, so there's no way really someone can instantiate it directly.

> Another benefit is that by moving the creation options into the constructor, the network instance can be re-used many times

This is potentially troublesome, as options contain a name, which should be unique.

**Comment by cristianrgreco** [comment]
Available in 2.10.0
